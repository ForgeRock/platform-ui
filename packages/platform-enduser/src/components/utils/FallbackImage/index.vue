<!-- Copyright (c) 2020-2021 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <div>
    <BImg
      v-if="imageFound"
      :src="src"
      :width="width"
      :height="width"
      :alt="alt" />
    <span
      v-else
      class="icon-holder p-2 d-flex bg-light border rounded">
      <FrIcon
        :class="['md-24 text-dark mt-auto', inputClass]"
        :name="fallback"
      />
    </span>
  </div>
</template>

<script>
import axios from 'axios';
import FrIcon from '@forgerock/platform-shared/src/components/Icon';

/**
 * @description Component that will load an image and if the image fails to load it will display a fallback image rather than loading nothing
 *
 * */
export default {
  name: 'ImageFallback',
  components: {
    FrIcon,
  },
  props: {
    inputClass: {
      type: String,
      default: () => '',
    },
    src: {
      type: String,
      default: () => '',
    },
    width: {
      type: [String, Number],
      default: () => '',
    },
    alt: {
      type: String,
      default: () => '',
    },
    fallback: {
      type: String,
      default: () => '',
    },
  },
  data() {
    return {
      imageFound: false,
    };
  },
  mounted() {
    if (this.src) {
      axios.get(`${this.src}`).then(({ status }) => {
        this.imageFound = status === 200;
      }).catch((error) => {
        if (error) {
          this.imageFound = false;
        }
      });
    } else {
      this.imageFound = false;
    }
  },
};
</script>

<style language="scss" scoped>
.icon-holder {
  min-height: 2.28rem;
  min-width: 2.28rem;
}
</style>
