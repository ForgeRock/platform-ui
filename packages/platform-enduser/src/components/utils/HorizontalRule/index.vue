<!-- Copyright 2019 ForgeRock AS. All Rights Reserved

Use of this code requires a commercial software license with ForgeRock AS.
or with one of its affiliates. All use shall be exclusively subject
to such license between the licensee and ForgeRock AS. -->
<template>
  <div class="row">
    <div class="col pr-0">
      <hr>
    </div>
    <div
      class="col-auto pt-1"
      v-html="insert">
      {{ insert }}
    </div>
    <div class="col pl-0">
      <hr>
    </div>
  </div>
</template>

<script>
/**
 * @description Component for making a horizontal rule line with text in the middle
 *
 * */
export default {
  name: 'HorizontalRule',
  props: {
    insert: {
      type: String,
      default: () => '',
    },
  },
};
</script>

<style lang="scss" scoped>
</style>
