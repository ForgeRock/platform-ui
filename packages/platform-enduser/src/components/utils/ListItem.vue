<template>
  <div
    v-if="collapsible"
    class="collapsible">
    <BListGroupItem
      href="#"
      :class="[{'list-item-cursor': collapsible===false}]"
      v-b-toggle="toggleId">
      <div class="media">
        <slot name="list-item-header" />
      </div>
    </BListGroupItem>

    <BCollapse
      :id="id"
      :visible="panelShown"
      @hide="$emit('hide')"
      @show="$emit('show')"
      @hidden="$emit('hidden')"
      @shown="$emit('shown')">
      <BCardBody class="pt-3">
        <slot name="list-item-collapse-body" />
      </BCardBody>
    </BCollapse>
  </div>

  <div
    v-else
    @click="$emit('row-click')"
    :class="[{'fr-hover-item': hoverItem}]">
    <BListGroupItem class="noncollapse">
      <div class="media">
        <slot name="list-item-header" />
      </div>
    </BListGroupItem>

    <BCardBody
      v-if="panelShown"
      class="pt-3">
      <slot name="list-item-collapse-body" />
    </BCardBody>
  </div>
</template>

<script>
/**
 * @description Used in conjunction with ListGroup.vue, this is the individual item in each list display.
 *
 * */
export default {
	name: 'ListItem',
	props: {
		collapsible: {
			type: Boolean,
			default: false,
		},
		panelShown: {
			type: Boolean,
			default: false,
		},
		hoverItem: {
			type: Boolean,
			default: false,
		},
	},
	data() {
		return {
			id: null,
		};
	},
	beforeMount() {
		// eslint-disable-next-line no-underscore-dangle
		this.id = `listItem${this._uid}`;
	},
	computed: {
		toggleId() {
			if (this.collapsible) {
				return this.id;
			}
			return null;
		},
	},
};
</script>
<style lang="scss" scoped>
    .media {
        height: 38px;

        & > * {
            -ms-flex-item-align: center;
            align-self: center;
        }
    }

    .list-item-cursor {
        cursor: default;
    }

    .fr-hover-item:hover {
        cursor: pointer;

        .list-group-item {
            background-color: $fr-hover-list-color;
        }
    }

    .collapsible:last-of-type > .list-group-item.collapsed {
        border-bottom-right-radius: $border-radius;
        border-bottom-left-radius: $border-radius;
        border-bottom: 1px solid $border-color;
    }

    .list-group-item-action {
        color: inherit;

        .caret {
            .caret-up {
                display: none;
            }
        }
        .meta {
            line-height: 1.2;
        }
        .btn-cancel {
            display: none;
        }
        .btn-edit {
            display: block;
        }

        &:not(.collapsed) {
            button {
                display: block;
            }
            .btn-cancel {
                display: block;
            }
            .btn-edit {
                display: none;
            }
            .caret {
                .caret-up {
                    display: block;
                }
                .caret-down {
                    display: none;
                }
            }
        }
    }

    .list-group-item-action:not(.collapsed),
    .list-group-item-action:not(.collapsed):hover,
    .list-group-item-action:not(.collapsed):focus,
    .noncollapse {
        background-color: $card-bg;
        border-bottom-color: transparent;
        cursor: initial;
    }

</style>
