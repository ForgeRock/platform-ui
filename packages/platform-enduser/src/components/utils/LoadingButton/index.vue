<!--
Copyright (c) 2020 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details.
-->

<template>
  <button
    :class="['btn btn-primary d-flex align-items-center', {'disabled': loading}]"
    type="button"
    @click="$emit('click')">
    <span :class="[{'align-self-start': !large, 'm-auto': large}]">
      {{ label }}
    </span>
    <span :class="['fr-grow', {'fr-grow-in': loading}]">
      <Transition name="fr-fade">
        <ClipLoader
          v-if="loading"
          :color="'white'"
          :size="'1rem'"
          class="position-relative fr-clip-loader ml-3" />
      </Transition>
    </span>
  </button>
</template>
<script>
import { ClipLoader } from 'vue-spinner/dist/vue-spinner.min';
/**
 * @description Button with built in loading spinner that can be used to show users something is loading
 *
 * */
export default {
  name: 'LoadingButton',
  components: { ClipLoader },
  props: {
    label: {
      type: String,
      default: () => '',
    },
    loading: {
      type: Boolean,
      default: false,
    },
    large: {
      type: Boolean,
      default: false,
    },
  },
};
</script>
<style lang="scss" scoped>
.fr-grow {
  width: 0;
  transition: all 0.25s ease;
}

.fr-grow-in {
  width: 3rem;
}

.fr-fade-enter-active {
  transition: all 0.125s ease;
}

.fr-fade-leave {
  transition: all 0.125s ease;
}

.fr-fade-enter,
.fr-fade-leave-to {
  opacity: 0;
}

.fr-fade-enter-to {
  opacity: 1;
}

.fr-clip-loader {
  top: 0.15rem;
}
</style>
