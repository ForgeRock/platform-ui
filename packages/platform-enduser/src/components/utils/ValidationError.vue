<template>
  <div
    v-show="validatorErrors.has(fieldName)"
    class="fr-validation-requirements text-left">
    <span class="text-danger">
      {{ validatorErrors.first(fieldName) }}
    </span>
  </div>
</template>

<script>
/**
 * @description Component for displaying error message for form fields
 *
 * */
export default {
  name: 'ValidationError',
  props: {
    validatorErrors: {
      type: Object,
      default: () => ({}),
    },
    fieldName: {
      type: String,
      default: () => '',
    },
  },
};
</script>

<style lang="scss" scoped></style>
