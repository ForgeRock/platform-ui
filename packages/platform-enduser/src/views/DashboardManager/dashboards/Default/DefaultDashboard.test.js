/**
 * Copyright (c) 2020-2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { mount, flushPromises } from '@vue/test-utils';
import { createStore } from 'vuex';
import Notifications from '@kyvg/vue3-notification';
import { setupTestPinia } from '@forgerock/platform-shared/src/utils/testPiniaHelpers';
import DefaultDashboard from './index';
import Welcome from '@/views/DashboardManager/dashboards/widgets/WelcomeWidget';

DefaultDashboard.mounted = jest.fn();

let wrapper;

describe('DefaultDashboard.vue', () => {
  const store = createStore({
    state: {
      SharedStore: { workforceEnabled: false, governanceEnabled: false },
    },
  });

  beforeEach(() => {
    jest.clearAllMocks();
    setupTestPinia();
    wrapper = mount(DefaultDashboard, {
      global: {
        plugins: [Notifications, store],
        mocks: {
          $t: (key) => (key),
          $router: { push: jest.fn() },
        },
      },
    });
  });

  describe('loading dashboard data', () => {
    it('executes loadWidgets method', async () => {
      const loadWidgetsSpy = jest.spyOn(wrapper.vm, 'loadWidgets');
      const widgetResponse = [{ type: 'Welcome' }];

      jest.spyOn(wrapper.vm, 'getRequestService').mockImplementation(() => ({
        get: () => Promise.resolve({
          data: {
            dashboard: {
              widgets: widgetResponse,
            },
          },
        }),
      }));

      wrapper.vm.loadWidgets();
      await flushPromises();

      expect(loadWidgetsSpy).toHaveBeenCalled();
      expect(wrapper.findComponent(Welcome).exists()).toBe(true);
      expect(wrapper.find('.my-applications-tiles').exists()).toBe(false);
      expect(wrapper.vm.widgets).toEqual([{ type: 'Welcome' }]);
    });

    it('error notification when getRequestService is not successful inside of loadWidgets', async () => {
      const notificationSpy = jest.spyOn(wrapper.vm, 'showErrorMessage');
      const error = { response: { data: { message: 'your widgets call failed' } } };

      jest.spyOn(wrapper.vm, 'getRequestService').mockImplementation(() => ({
        get: () => Promise.reject(error),
      }));

      wrapper.vm.loadWidgets();
      await flushPromises();
      expect(notificationSpy).toHaveBeenCalledWith(error, 'pages.dashboard.errorGetApplications');
    });

    it('executes loadConsumerApplications method', async () => {
      const loadApplicationsSpy = jest.spyOn(wrapper.vm, 'loadConsumerApplications');

      jest.spyOn(wrapper.vm, 'getRequestService').mockImplementation(() => ({
        get: () => Promise.resolve({
          data: [
            { dashboardDisplayName: 'b', dashboardLogin: ['test'], dashboardIcon: ['test'] },
            { dashboardDisplayName: 'c', dashboardLogin: ['test'], dashboardIcon: ['test'] },
            { dashboardDisplayName: 'a', dashboardLogin: ['test'], dashboardIcon: ['test'] },
          ],
        }),
      }));

      wrapper.vm.loadConsumerApplications();
      await flushPromises();
      expect(loadApplicationsSpy).toHaveBeenCalled();
      expect(wrapper.find('.list-group').exists()).toBe(true);

      // sorted by method
      expect(wrapper.vm.myApplications).toStrictEqual([
        { dashboardDisplayName: 'a', dashboardLogin: ['test'], dashboardIcon: ['test'] },
        { dashboardDisplayName: 'b', dashboardLogin: ['test'], dashboardIcon: ['test'] },
        { dashboardDisplayName: 'c', dashboardLogin: ['test'], dashboardIcon: ['test'] },
      ]);
    });

    it('error notification when getRequestService is not successful within loadConsumerApplications', async () => {
      const notificationSpy = jest.spyOn(wrapper.vm, 'showErrorMessage');
      const error = 'your applications call failed';

      jest.spyOn(wrapper.vm, 'getRequestService').mockImplementation(() => ({
        get: () => Promise.reject(error),
      }));

      wrapper.vm.loadConsumerApplications();
      await flushPromises();
      expect(notificationSpy).toHaveBeenCalledWith('your applications call failed', 'pages.dashboard.errorGetApplications');
    });
  });
});

describe('cloud', () => {
  const store = createStore({
    state: {
      workflow: true,
      isFraas: true,
    },
  });

  it('hides the workflow widget when isFraas is true', async () => {
    wrapper = mount(DefaultDashboard, {
      global: {
        plugins: [Notifications, store],
        mocks: {
          $t: (key) => (key),
        },
      },
    });

    const widgetResponse = [{ type: 'Welcome' }];

    jest.spyOn(wrapper.vm, 'getRequestService').mockImplementation(() => ({
      get: () => Promise.resolve({
        data: {
          dashboard: {
            widgets: widgetResponse,
          },
        },
      }),
    }));

    wrapper.vm.loadWidgets();
    await flushPromises();
    expect(wrapper.vm.widgets).toEqual([{ type: 'Welcome' }]);
  });
});
