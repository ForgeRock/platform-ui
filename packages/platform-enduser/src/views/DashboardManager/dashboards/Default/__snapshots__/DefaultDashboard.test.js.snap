// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultDashboard.vue loading dashboard data executes loadConsumerApplications method: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (6)",
    "
",
    " at <ListGroup",
    "title="pages.dashboard.applications.heading"",
    ">",
    "
",
    " at <BCol>",
    "
",
    " at <BContainer",
    "key=0",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) (5)",
    "
",
    " at <ListGroup",
    "title="pages.dashboard.applications.heading"",
    ">",
    "
",
    " at <BCol>",
    "
",
    " at <BContainer",
    "key=0",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (6)",
    "
",
    " at <BCard",
    "no-body=""",
    "class="mb-3"",
    ">",
    "
",
    " at <ListGroup",
    "title="pages.dashboard.applications.heading"",
    ">",
    "
",
    " at <BCol>",
    "
",
    " at <BContainer",
    "key=0",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (7)",
    "
",
    " at <BCardBody",
    "key=0",
    "class="py-4 list-group-header-card"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    "class="mb-3"",
    ">",
    "
",
    " at <ListGroup",
    "title="pages.dashboard.applications.heading"",
    ">",
    "
",
    " at <BCol>",
    "
",
    " at <BContainer",
    "key=0",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (8)",
    "
",
    " at <BListGroup",
    "flush=""",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    "class="mb-3"",
    ">",
    "
",
    " at <ListGroup",
    "title="pages.dashboard.applications.heading"",
    ">",
    "
",
    " at <BCol>",
    "
",
    " at <BContainer",
    "key=0",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (7)",
    "
",
    " at <ListItem",
    "collapsible=false",
    ">",
    "
",
    " at <ConsumerApplications",
    "key="a"",
    "application-details=",
    {
      "dashboardDisplayName": "a",
      "dashboardIcon": [
        "test",
      ],
      "dashboardLogin": [
        "test",
      ],
    },
    ">",
    "
",
    " at <BListGroup",
    "flush=""",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    "class="mb-3"",
    ">",
    "
",
    " at <ListGroup",
    "title="pages.dashboard.applications.heading"",
    ">",
    "
",
    " at <BCol>",
    "
",
    " at <BContainer",
    "key=0",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) (6)",
    "
",
    " at <ListItem",
    "collapsible=false",
    ">",
    "
",
    " at <ConsumerApplications",
    "key="a"",
    "application-details=",
    {
      "dashboardDisplayName": "a",
      "dashboardIcon": [
        "test",
      ],
      "dashboardLogin": [
        "test",
      ],
    },
    ">",
    "
",
    " at <BListGroup",
    "flush=""",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    "class="mb-3"",
    ">",
    "
",
    " at <ListGroup",
    "title="pages.dashboard.applications.heading"",
    ">",
    "
",
    " at <BCol>",
    "
",
    " at <BContainer",
    "key=0",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (9)",
    "
",
    " at <BListGroupItem",
    "class="noncollapse"",
    ">",
    "
",
    " at <ListItem",
    "collapsible=false",
    ">",
    "
",
    " at <ConsumerApplications",
    "key="a"",
    "application-details=",
    {
      "dashboardDisplayName": "a",
      "dashboardIcon": [
        "test",
      ],
      "dashboardLogin": [
        "test",
      ],
    },
    ">",
    "
",
    " at <BListGroup",
    "flush=""",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    "class="mb-3"",
    ">",
    "
",
    " at <ListGroup",
    "title="pages.dashboard.applications.heading"",
    ">",
    "
",
    " at <BCol>",
    "
",
    " at <BContainer",
    "key=0",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: Failed to resolve component: BImg
If this is a native custom element, make sure to exclude it from component resolution via compilerOptions.isCustomElement.",
    "
",
    " at <ImageFallback",
    "fallback="exit_to_app"",
    "src="test"",
    "alt="pages.dashboard.applications.listItemBrandIconAltText"",
    ">",
    "
",
    " at <BListGroupItem",
    "class="noncollapse"",
    ">",
    "
",
    " at <ListItem",
    "collapsible=false",
    ">",
    "
",
    " at <ConsumerApplications",
    "key="a"",
    "application-details=",
    {
      "dashboardDisplayName": "a",
      "dashboardIcon": [
        "test",
      ],
      "dashboardLogin": [
        "test",
      ],
    },
    ">",
    "
",
    " at <BListGroup",
    "flush=""",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    "class="mb-3"",
    ">",
    "
",
    " at <ListGroup",
    "title="pages.dashboard.applications.heading"",
    ">",
    "
",
    " at <BCol>",
    "
",
    " at <BContainer",
    "key=0",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: Failed to resolve component: BImg
If this is a native custom element, make sure to exclude it from component resolution via compilerOptions.isCustomElement.",
    "
",
    " at <ImageFallback",
    "fallback="exit_to_app"",
    "src="test"",
    "alt="pages.dashboard.applications.listItemBrandIconAltText"",
    ">",
    "
",
    " at <BListGroupItem",
    "class="noncollapse"",
    ">",
    "
",
    " at <ListItem",
    "collapsible=false",
    ">",
    "
",
    " at <ConsumerApplications",
    "key="b"",
    "application-details=",
    {
      "dashboardDisplayName": "b",
      "dashboardIcon": [
        "test",
      ],
      "dashboardLogin": [
        "test",
      ],
    },
    ">",
    "
",
    " at <BListGroup",
    "flush=""",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    "class="mb-3"",
    ">",
    "
",
    " at <ListGroup",
    "title="pages.dashboard.applications.heading"",
    ">",
    "
",
    " at <BCol>",
    "
",
    " at <BContainer",
    "key=0",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: Failed to resolve component: BImg
If this is a native custom element, make sure to exclude it from component resolution via compilerOptions.isCustomElement.",
    "
",
    " at <ImageFallback",
    "fallback="exit_to_app"",
    "src="test"",
    "alt="pages.dashboard.applications.listItemBrandIconAltText"",
    ">",
    "
",
    " at <BListGroupItem",
    "class="noncollapse"",
    ">",
    "
",
    " at <ListItem",
    "collapsible=false",
    ">",
    "
",
    " at <ConsumerApplications",
    "key="c"",
    "application-details=",
    {
      "dashboardDisplayName": "c",
      "dashboardIcon": [
        "test",
      ],
      "dashboardLogin": [
        "test",
      ],
    },
    ">",
    "
",
    " at <BListGroup",
    "flush=""",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    "class="mb-3"",
    ">",
    "
",
    " at <ListGroup",
    "title="pages.dashboard.applications.heading"",
    ">",
    "
",
    " at <BCol>",
    "
",
    " at <BContainer",
    "key=0",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`DefaultDashboard.vue loading dashboard data executes loadWidgets method: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) Vue 3's render function API has changed. You can opt-in to the new API with:

  configureCompat({ RENDER_FUNCTION: false })

  (This can also be done per-component via the "compatConfig" option.)
  Details: https://v3-migration.vuejs.org/breaking-changes/render-function-api.html",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (2)",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) Functional component <BContainer> should be defined as a plain function in Vue 3. The "functional" option has been removed. NOTE: Before migrating to use plain functions for functional components, first make sure that all async components usage have been migrated and its compat behavior has been disabled.
  Details: https://v3-migration.vuejs.org/breaking-changes/functional-components.html",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) vm.$listeners has been removed. In Vue 3, parent v-on listeners are included in vm.$attrs and it is no longer necessary to separately use v-on="$listeners" if you are already using v-bind="$attrs". (Note: the Vue 3 behavior only applies if this compat config is disabled)
  Details: https://v3-migration.vuejs.org/breaking-changes/listeners-removed.html",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (2)",
    "
",
    " at <BJumbotron",
    "key=1",
    "class="mt-4 text-center"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (3)",
    "
",
    " at <WelcomeWidget",
    "details=undefined",
    "display-compact-header=false",
    ">",
    "
",
    " at <BRow",
    "key=0",
    "class="mx-4"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) (2)",
    "
",
    " at <WelcomeWidget",
    "details=undefined",
    "display-compact-header=false",
    ">",
    "
",
    " at <BRow",
    "key=0",
    "class="mx-4"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: Property "$pinia" was accessed during render but is not defined on instance.",
    "
",
    " at <BJumbotron",
    "class="text-center"",
    ">",
    "
",
    " at <WelcomeWidget",
    "details=undefined",
    "display-compact-header=false",
    ">",
    "
",
    " at <BRow",
    "key=0",
    "class="mx-4"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: Property "$pinia" was accessed during render but is not defined on instance.",
    "
",
    " at <BJumbotron",
    "class="text-center"",
    ">",
    "
",
    " at <WelcomeWidget",
    "details=undefined",
    "display-compact-header=false",
    ">",
    "
",
    " at <BRow",
    "key=0",
    "class="mx-4"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (2)",
    "
",
    " at <BAvatar",
    "variant="link"",
    "size="112px"",
    "src=""",
    ">",
    "
",
    " at <BJumbotron",
    "class="text-center"",
    ">",
    "
",
    " at <WelcomeWidget",
    "details=undefined",
    "display-compact-header=false",
    ">",
    "
",
    " at <BRow",
    "key=0",
    "class="mx-4"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_SCOPED_SLOTS) vm.$scopedSlots has been removed. Use vm.$slots instead.
  Details: https://v3-migration.vuejs.org/breaking-changes/slots-unification.html",
    "
",
    " at <BAvatar",
    "variant="link"",
    "size="112px"",
    "src=""",
    ">",
    "
",
    " at <BJumbotron",
    "class="text-center"",
    ">",
    "
",
    " at <WelcomeWidget",
    "details=undefined",
    "display-compact-header=false",
    ">",
    "
",
    " at <BRow",
    "key=0",
    "class="mx-4"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (4)",
    "
",
    " at <BAvatar",
    "variant="link"",
    "size="112px"",
    "src=""",
    ">",
    "
",
    " at <BJumbotron",
    "class="text-center"",
    ">",
    "
",
    " at <WelcomeWidget",
    "details=undefined",
    "display-compact-header=false",
    ">",
    "
",
    " at <BRow",
    "key=0",
    "class="mx-4"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) (3)",
    "
",
    " at <BAvatar",
    "variant="link"",
    "size="112px"",
    "src=""",
    ">",
    "
",
    " at <BJumbotron",
    "class="text-center"",
    ">",
    "
",
    " at <WelcomeWidget",
    "details=undefined",
    "display-compact-header=false",
    ">",
    "
",
    " at <BRow",
    "key=0",
    "class="mx-4"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (3)",
    "
",
    " at <BIconPersonFill",
    "aria-hidden="true"",
    "alt="avatar"",
    ">",
    "
",
    " at <BAvatar",
    "variant="link"",
    "size="112px"",
    "src=""",
    ">",
    "
",
    " at <BJumbotron",
    "class="text-center"",
    ">",
    "
",
    " at <WelcomeWidget",
    "details=undefined",
    "display-compact-header=false",
    ">",
    "
",
    " at <BRow",
    "key=0",
    "class="mx-4"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (5)",
    "
",
    " at <BIconPersonFill",
    "aria-hidden="true"",
    "alt="avatar"",
    ">",
    "
",
    " at <BAvatar",
    "variant="link"",
    "size="112px"",
    "src=""",
    ">",
    "
",
    " at <BJumbotron",
    "class="text-center"",
    ">",
    "
",
    " at <WelcomeWidget",
    "details=undefined",
    "display-compact-header=false",
    ">",
    "
",
    " at <BRow",
    "key=0",
    "class="mx-4"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) (4)",
    "
",
    " at <BIconPersonFill",
    "aria-hidden="true"",
    "alt="avatar"",
    ">",
    "
",
    " at <BAvatar",
    "variant="link"",
    "size="112px"",
    "src=""",
    ">",
    "
",
    " at <BJumbotron",
    "class="text-center"",
    ">",
    "
",
    " at <WelcomeWidget",
    "details=undefined",
    "display-compact-header=false",
    ">",
    "
",
    " at <BRow",
    "key=0",
    "class="mx-4"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (4)",
    "
",
    " at <BIconBase",
    "aria-label="person fill"",
    "aria-hidden="true"",
    "alt="avatar"",
    " ...",
    ">",
    "
",
    " at <BIconPersonFill",
    "aria-hidden="true"",
    "alt="avatar"",
    ">",
    "
",
    " at <BAvatar",
    "variant="link"",
    "size="112px"",
    "src=""",
    ">",
    "
",
    " at <BJumbotron",
    "class="text-center"",
    ">",
    "
",
    " at <WelcomeWidget",
    "details=undefined",
    "display-compact-header=false",
    ">",
    "
",
    " at <BRow",
    "key=0",
    "class="mx-4"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (5)",
    "
",
    " at <BButton",
    "onClick=fn",
    "variant="primary"",
    "class="mt-2"",
    ">",
    "
",
    " at <BJumbotron",
    "class="text-center"",
    ">",
    "
",
    " at <WelcomeWidget",
    "details=undefined",
    "display-compact-header=false",
    ">",
    "
",
    " at <BRow",
    "key=0",
    "class="mx-4"",
    ">",
    "
",
    " at <BContainer",
    "fluid="true"",
    ">",
    "
",
    " at <DefaultDashboard",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;
